vm/systemSmartContracts/staking.go

```diff
@@ -0,0 +1,226 @@
+			log.Error("unmarshal error on staking smart contract stake function " + err.Error())
+			return vmcommon.UserError
+		}
+	}
+
+	if registrationData.Staked == true {
+		log.Error("account already staked, re-staking is invalid")
+		return vmcommon.UserError
+	}
+
+	registrationData.Staked = true
+
+	if len(args.Arguments) < 1 {
+		log.Error("not enough arguments to process stake function")
+		return vmcommon.UserError
+	}
+
+	registrationData.StartNonce = args.Header.Number.Uint64()
+	registrationData.BlsPubKey = args.Arguments[0].Bytes()
+	//TODO: verify if blsPubKey is valid
+
+	data, err := json.Marshal(registrationData)
+	if err != nil {
+		log.Error("marshal error on staking smart contract stake function " + err.Error())
+		return vmcommon.UserError
+	}
+
+	r.eei.SetStorage(args.CallerAddr, data)
+
+	err = r.eei.Transfer(args.RecipientAddr, args.CallerAddr, args.CallValue, nil)
+	if err != nil {
+		log.Error("transfer error on stake function " + err.Error())
+	}
+
+	return vmcommon.Ok
+}
+
+func (r *stakingSC) unStake(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	var registrationData stakingData
+	data := r.eei.GetStorage(args.CallerAddr)
+	if data == nil {
+		log.Error("unStake is not possible for address which is not staked")
+		return vmcommon.UserError
+	}
+
+	err := json.Unmarshal(data, registrationData)
+	if err != nil {
+		log.Error("unmarshal error in unStake function of staking smart contract " + err.Error())
+		return vmcommon.UserError
+	}
+
+	registrationData.Staked = false
+	registrationData.UnStakedNonce = args.Header.Number.Uint64()
+
+	data, err = json.Marshal(registrationData)
+	if err != nil {
+		log.Error("marshal error in unStake function of staking smart contract" + err.Error())
+		return vmcommon.UserError
+	}
+
+	r.eei.SetStorage(args.CallerAddr, data)
+
+	return vmcommon.Ok
+}
+
+func (r *stakingSC) finalizeUnStake(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	ownerAddress := r.eei.GetStorage([]byte(ownerKey))
+	if !bytes.Equal(ownerAddress, args.CallerAddr) {
+		return vmcommon.UserError
+	}
+
+	var registrationData stakingData
+	for _, arg := range args.Arguments {
+		data := r.eei.GetStorage(arg.Bytes())
+		err := json.Unmarshal(data, registrationData)
+		if err != nil {
+			log.Error("unmarshal error on finalize unstake function" + err.Error())
+			return vmcommon.UserError
+		}
+
+		if registrationData.UnStakedNonce == 0 {
+			log.Error("validator did not unstaked yet")
+			return vmcommon.UserError
+		}
+
+		r.eei.SetStorage(arg.Bytes(), nil)
+
+		err = r.eei.Transfer(args.CallerAddr, arg.Bytes(), registrationData.StakeValue, nil)
+		if err != nil {
+			log.Error("transfer error on finalizeUnStake function " + err.Error())
+			return vmcommon.UserError
+		}
+	}
+	return vmcommon.Ok
+}
+
+func (r *stakingSC) slash(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	ownerAddress := r.eei.GetStorage([]byte(ownerKey))
+	if !bytes.Equal(ownerAddress, args.CallerAddr) {
```

Comment:
I do not think slash can work like this. Who is the owner that can slash one node?

Reply:
The owner of the smart contract is predefined, so any metachain node can do it. But the block is only signed when there is no wrong slashing. Otherwise people might just call from the own addresses the slashing function, and you do not know at the VM level if it was a good call or not. Deciding whether it is a good call or not it is part of peer processor, block creation.